### ISSUE TITLE
[2.x] Flowable.parallel() consumes and produces excessive elements.

### ISSUE THREAD

**Issue by daemas** [ReactiveX/RxJava#6394]
Flowable.parallel() produces more elements than requested and consumes even more, even if `prefetch` is set to 1.

Sample:
```java
import io.reactivex.Maybe;
import io.reactivex.schedulers.Schedulers;

import java.util.concurrent.atomic.AtomicInteger;

public class TestPrefetch {
    public static void main(String[] args) {
        AtomicInteger i = new AtomicInteger(0);
        Maybe<Integer> maybeSource = Maybe.create(emitter -> emitter.onSuccess(i.incrementAndGet()));
        maybeSource
                .repeat()
                .parallel(30, 1)
                .runOn(Schedulers.io(), 1)
                .map(Object::toString)
                .sequential(1)
                .take(100)
                .blockingSubscribe(System.out::println);
        System.out.println("Consumed: " + i.get());
    }
}
```

Actual output:
```plain
1
2
3
...
104
Consumed: 145
```

Expected output:
```plain
1
2
...
100
Consumed: 100
```

Tested on versions 2.2.5 and 2.2.6.

**Comment by akarnokd** [comment]
This behavior is acceptable. There are two effects in play here:

1) The parallelism introduces an asynchrony between the decision to stop the sequence and the requesting for more items. 29 of the 30 rails in the sequence can go ahead more before they get eventually cancelled by the 1 rail servicing `take()`. Try with different parallelism level and see it will produce a bit more.

2) If you index the result values you'll see that you get 100 items, but their order can be arbitrary as it may get values from rails that went above 100 due to reason 1). Try this and you see there are no more than 100 item printouts:

```java
AtomicInteger j = new AtomicInteger(0);
AtomicInteger i = new AtomicInteger(0);
Maybe<Integer> maybeSource = Maybe.create(emitter -> emitter.onSuccess(i.incrementAndGet()));
maybeSource
.repeat()
.parallel(30, 1)
.runOn(Schedulers.io(), 1)
.map(Object::toString)
.sequential(1)
.take(100)
.blockingSubscribe(v -> System.out.println(j.incrementAndGet() + " - " + v));

System.out.println("Consumed: " + i.get());
```

**Comment by daemas** [comment]
Thank you for your answer @akarnokd.
1. So is there currently no option to fetch from the source the exact amount of items requested by the subscriber and process them in parallel? I tried to do it the "old-fashioned" way with flatMap(), but it also seem to fetch a little bit more than requested.
2. Oh, sure, how could I overlook the fact that the order of items is not guaranteed if they are processed in parallel.

**Comment by akarnokd** [comment]
There is no way to use the backpressure across multiple operators and async stages to request from a source. Use a feedback loop with `PublishProcessor` where you trigger requests whenever you call onNext regardless of the backpressure status of other stages.

**Comment by daemas** [comment]
> There is no way to use the backpressure across multiple operators and async stages to request from a source.

Well, I'm trying to understand why, and I'm failing to. I think of a reactive pull-based pipeline. To simplify, let's imagine there are only a producer and a consumer of the items. The consumer asks for (pulls) exactly one item, processes it, and after that asks for another one. The producer does not emit the next item unless the consumer asks for it. Right? If we introduce an intermediate stage, it becomes a producer for the last stage and a consumer for the first one. So when the last stage requests an item the intermediate stage just passes the request to the first stage, and then receives and processes the item and passes the result to the last stage. Then why in my example the very first producer (`maybeSource`) produces more items then requested by the very last consumer? Who requests them in the chain of operators, and why?

**Comment by akarnokd** [comment]
Don't think of it as one entity. A flow is assembled from multiple stages and backpressure is negotiated on a stage boundary where individual stages can decide on the request pattern themselves. Some don't interfere, some batch and some keep prefetching when the downstream's `onNext` returns. This is a non-trivial topic and requires a lot of experience to be able to fully understand the consequences. Many times, 1:1 is not even clear to begin with, i.e., when merging two async sources, requesting 1 should request an item from which source?

**Comment by daemas** [comment]
> Some don't interfere, some batch and some keep prefetching when the downstream's `onNext` returns.

Well, many operators accept `bufferSize` argument, so I supposed I would be able to control the prefetching and batching behaviour with it's help. Looks like it's not the case.

> This is a non-trivial topic and requires a lot of experience to be able to fully understand the consequences.

Indeed, I am not too experienced with the RX library. I came from Scala's `Future` world, where there are no means of controlling backpressure "out of the box", so I had to control and fine-tune it manually in the places where it was necessary. Hence is my surprise to find I cannot fine-tune it the same way with RX, even if I specify mentioned `bufferSize` parameters.

> Many times, 1:1 is not even clear to begin with, i.e., when merging two async sources, requesting 1 should request an item from which source?

Well, there are several commonly used balancing scenarios. Round-robin is my favorite for it's simplicity and general applicability :) Also users might choose one of pre-defined strategies or implement and provide a custom one.

Anyway, thank you for explaining all these aspects to me. Like I said, I'm new to the RX world, and just getting used to it. So I feel I'm just trying to use my old habits within the new environment, and I have to find another way.
